File: robot/src/jade/JadeDeserializer.java
package jade;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

/**
 * The JadeDeserializer is used to convert the contents of a file or a string to a new JadeObject.
 * <p> 
 * Structure of an example .jde file:
 * <p>
 * <i>variableOne: 'valueOne';</i>
 * <p>
 * <i>variableTwo: 'valueTwo';</i>
 * <p>
 * <i>:</i> -> used to split the variable and its value.
 * <p>
 * <i>;</i> -> to denote the end of a varialbe.
 * <p>
 * <i>'</i> -> surrounds a value of a variable.
 *
 * @author  Joseph Beck
 * @version 0.1
 * @since   2023
 */
public class JadeDeserializer {

    /**
     * Empty JadeDeserializer constructor, it has no variables that requitre initialization.
     * 
     * @return none
     */
    public JadeDeserializer() { }

    /**
     * Deserializes a string from a given file into a JadeObject.
     * The file and its strings require a large amount of parsing for correct JadeObject creation.
     * New lines are removed, data is split based on ; and pairs are determined by : and ''.
     * 
     * @param String directory
     * @return JadeObject
     */
    public JadeObject deserializeFromFile(String dir) {
        try (BufferedReader reader = new BufferedReader(new FileReader(dir))) {
            String line = reader.readLine();
            String data = "";

            while (line != null) {
                data += line;
                line = reader.readLine();
            }
            
            data.replaceAll("\\s+" , "");
            String[] elements = data.split(";");
            JadePair[] dataPoints = new JadePair[elements.length];

            for (int i = 0; i < elements.length; i++) {
                String[] tuple = elements[i].split(":");
                dataPoints[i] = makeJadePair(tuple);
            }
            
            JadeObject obj = new JadeObject(dataPoints);
            return obj;
        } catch (IOException e) {
            e.printStackTrace();
            return null;   
        }
    }

    /**
     * Deserializes a given string.
     * The string itself is first parsed removeing any unnceresary artifcats.
     * It is split identically to the file splitting.
     * 
     * @param String jade string
     * @return JadeObject
     */
    public JadeObject deserializeFromString(String jadeString) {
        jadeString.replaceAll("\\s+" , "");
        String[] elements = jadeString.split(";");
        JadePair[] dataPoints = new JadePair[elements.length];
        
        for (int i = 0; i < elements.length; i++) {
            String[] tuple = elements[i].split(":");
            dataPoints[i] = makeJadePair(tuple);
        }
        
        JadeObject obj = new JadeObject(dataPoints);
        return obj;
    }

    /**
     * Creates a JadePair from a string array of elements that should be of length 2.
     * Each element is also correctly parsed to ensure there are no left over artifacts.
     * 
     * @param String[] elements
     * @return JadePair
     */
    private JadePair makeJadePair(String[] elements) {
        String first = elements[0]
            .replaceAll("\"", "")
            .replaceAll("\\s+", "")
            .replaceAll("'", "");

        String second = elements[1]
            .replaceAll("\"", "")
            .replaceAll("\\s+", "")
            .replaceAll("'", "");
    
        return new JadePair(first, second);
    }
}

File: client/jade/jade_deserializer.rb
# frozen_string_literal: true

require './jade/jade_pair'
require './jade/jade_object'

# The JadeDeserializer converts either a string or .jde file that is formatted correctly to a JadeObject.
# It does this by parsing the data and creating a series of JadePairs which are used as the arguments of the JadeObject.
#
# Formatting (with pretty printing):
#
# <variableOne>: '<valueOne>';
#
# <variableTwo>: '<valueTwo>';
#
# Formatting (without pretty printing):
#
# <variableOne>: '<valueOne>'; <variableTwo>: '<valueTwo>';
#
# : -> Splits the variable and its value
#
# ; -> Denotes an end of line
#
# ' -> Surround the value of the variable
#
# " -> Can also surround the value of the variable
#
# When parsing Jade ignores ALL whitespace, even in variables.
#
# @author Joseph Beck
# @!visibility public
class JadeDeserializer

  public

  # The from_str method creates a JadeObject from a given string, as long as it able to be converted to JadeObject.
  # It splits it into pairs with .split(';').
  # All new lines are also removed at this point.
  # The pairs are then parsed in another function.
  #
  # @param [String]
  # @return [JadeObject]
  # @!visibility public
  def from_str(jaded_str)
    data = jaded_str.tr("\n", '').split(';')
    make_jade_object(data)
  end

  # The from_file method creates a JadeObject from a given file directory, as long as it able to be converted to JadeObject.
  # It first reads in the file and then splits it into pairs with .split(';').
  # All new line are also removed at this point.
  # The pairs are then parsed in another function.
  #
  # @param [String]
  # @return [JadeObject]
  # @!visibility public
  def from_file(dir)
    file = File.open(dir)
    data = file.read.tr("\n", '').split(';')
    make_jade_object(data)
  end

  private

  # The make_jade_object takes in a half-parsed array of strings and coverts these into JadePairs.
  # It does this for every piece of data in the array.
  # During this all whitespace is removed and it is split but : to get each pair.
  # Each created pair is appended to the jade_pairs array.
  # This array is later used as args for the new JadeObject which is returned.
  #
  # @param [String[]]
  # @return [JadeObject]
  # @!visibility public
  def make_jade_object(data)
    jade_pairs = []

    data.each do |piece|
      piece = piece.tr(' ', '').tr("'", '')
      first, second = piece.split(':')
      first = first.tr(' ', '').tr("\n", '')
      second = second.tr(' ', '').tr("\n", '')
      jade_pairs << JadePair.new(first, second)
    end
    JadeObject.new(jade_pairs)
  end
end
